docs/memory_layout.md

```diff
@@ -0,0 +1,300 @@
+| mat3x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]</kbd> |
+| mat4x2\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]<br/>[bbbb][bbbb][bbbb][bbbb]</kbd> |
+| mat4x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]</kbd> |
+
+## Structures
+```
+struct {
+  int32 w;
+  float x;
+}
+```
+
+The STD140 pads 8 bytes at the end to become a multiple of 16 bytes.
+
+| STD | Array Stride | Bytes |
+|-----|:--------------------:|-------|
+| 140 | 16 | <kbd>{w [bbbb]}<br/>{x [bbbb]}<br/> [----][----]</kbd> |
+| 430 |  8 | <kbd>{w [bbbb]}<br/>{x [bbbb]}</kbd> |
+
+<hr>
+```
+struct {
+  struct {
+    int32 a;
+    float b;
+  } x;
+  float y;
+}
+```
+
+The STD140 pads 8 bytes at the end to become a multiple of 16 bytes.
+
+| STD | Array Stride | Bytes |
+|-----|:--------------------:|-------|
+| 140 | 32 | <kbd>{x {a [bbbb]}<br/> &nbsp;&nbsp; {b [bbbb]}<br/> &nbsp;&nbsp; [----][----]<br/>{y [bbbb][----][----][----]}</kbd> |
+| 430 | 12 | <kbd>{x {a [bbbb]}<br/> &nbsp;&nbsp; {b [bbbb]}<br/>{y [bbbb]}</kbd> |
+
+
+<hr>
+
+```
+struct {
+  int32 w;
+  vec2<float> x;
+  float y;
+}
+```
+
+For both cases, the `vec2` is the member with the largest base alignment
+requirement, so giving a base alignment of 8 bytes. The STD140 cases pads 8
+bytes at the end (in the array element case) to become a multiple of 16.
+
+| STD | Array Stride  | Bytes |
+|-----|:---------------------:|-------|
+| 140 | 32 | <kbd>{w [bbbb][----]}<br/>{x [bbbb][bbbb]}<br/>{y [bbbb][----]}<br/> [----][----]</kbd> |
+| 430 | 24 | <kbd>{w [bbbb][----]}<br/>{x [bbbb][bbbb]}<br/>{y [bbbb][----]}</kbd> |
+
+<hr>
+
+
+```
+struct {
+  int32 w;
+  vec3<float> x;
+  float y;
+}
+```
+
+The `vec3` expands to a `vec4`. This gives a base alignment of 16 bytes. So
+the `w` pads to 16 bytes. the `float y` is packed into the `vec3` as there is
+space (effectively making it a vec4).
+
+
+| STD | Array Stride | Bytes |
+|-----|:--------------------:|-------|
+| 140 | 32 | <kbd>{w [bbbb][----][----][----]}<br/>{x [bbbb][bbbb][bbbb]}<br/>{y [bbbb]}</kbd> |
+| 430 | 32 | <kbd>{w [bbbb][----][----][----]}<br/>{x [bbbb][bbbb][bbbb]}<br/>{y [bbbb]}</kbd> |
+
+<hr>
+
+
+```
+struct {
+  int32 w;
+  mat2x2<float> x;
+  float y;
+}
+```
+
+In STD140 the `mat2x2<float>` has a base alignment of 16 bytes (the mat2x2 is, effectively,
+an array of vec2's The vec2's have a size of 8 bytes this size rounds up to a vec4, so 16 bytes).
+
+In STD430, the round up doesn't happen, so the base alignment is 8 bytes.
+
+| STD | Array Stride  | Bytes |
+|-----|:---------------------:|-------|
+| 140 | 64 | <kbd>{w [bbbb][----][----][----]}<br/>{x [bbbb][bbbb][----][----]<br/> &nbsp;&nbsp; [bbbb][bbbb][----][----]}<br/>{y [bbbb][----][----][----]}</kbd> |
+| 430 | 48 | <kbd>{w [bbbb][----]}<br/>{x [bbbb][bbbb][bbbb][bbbb]}<br/>{y [bbbb][----]}</kbd> |
```

Comment:
Isn't the array stride for the 430 case only 32?

Reply:
yes, yes it is.